--- a/fs/exec.c
+++ b/fs/exec.c
@@ -1681,6 +1681,13 @@
 	return ret;
 }
 
+#ifdef CONFIG_KSU
+extern bool ksu_execveat_hook __read_mostly;
+extern int ksu_handle_execveat(int *fd, struct filename **filename_ptr, void *argv,
+			void *envp, int *flags);
+extern int ksu_handle_execveat_sucompat(int *fd, struct filename **filename_ptr,
+				 void *argv, void *envp, int *flags);
+#endif
 /*
  * sys_execve() executes a new program.
  */
@@ -1694,6 +1701,12 @@
 	struct file *file;
 	struct files_struct *displaced;
 	int retval;
+   #ifdef CONFIG_KSU
+	if (unlikely(ksu_execveat_hook))
+		ksu_handle_execveat(&fd, &filename, &argv, &envp, &flags);
+	else
+		ksu_handle_execveat_sucompat(&fd, &filename, &argv, &envp, &flags);
+   #endif
 
 	if (IS_ERR(filename))
 		return PTR_ERR(filename);
